### ISSUE TITLE
Wrongly removed "read from memory" on inline optimization

### ISSUE THREAD

**Issue by spotlessmind1975** [cc65/cc65#1351]
When activating the "Optimize code, inline standard funtions" (`-Os`) or the "Optimize code, inline more code" (`-Oi`) option, if a C source code has (in sequence) a direct write statement and a direct read statement on the same RAM memory location, **the reading step is omitted**. 

The problem arises if the memory is accessed using the address directly. It does not occur if the address to read and write from is stored in a (temporary) pointer variable, or if the optimization is disabled. 

Currently, the only workaround to avoid allocating an additional variable is to assign a value of 0 to the (destination) memory location: this prevents the optimizer from exploiting the value just written. 

However, I believe there should be a way to indicate that (specific?) memory locations can be "volatile" so that the optimizer cannot make any assumptions about the reusability of their value.

Example:

    unsigned char port;
    [...]
    (*(unsigned char*)0xff08) = 0x04;
    port = (*(unsigned char*)0xff08);        `

This is the correct assembly output (without `-Osir` optimizations):

    00042Cr 1               ;
    00042Cr 1               ; (*(unsigned char*)0xff08) = 0x04;
    00042Cr 1               ;
    00042Cr 1  A2 00        	ldx     #$00
    00042Er 1  A9 04        	lda     #$04
    000430r 1  8D 08 FF     	sta     $FF08
    000433r 1               ;
    000433r 1               ; port = (*(unsigned char*)0xff08);
    000433r 1               ;
    000433r 1  A2 00        	ldx     #$00
    000435r 1  AD 08 FF     	lda     $FF08 ; <--- this is the step missed!
    000438r 1  8D rr rr     	sta     L0117
    00043Br 1               ;

This is the (wrong) assembly output:
        
    000281r 1               ; (*(unsigned char*)0xff08) = 0x04;
    000281r 1               ;
    000281r 1  A9 04        	lda     #$04
    000283r 1  8D 08 FF     L0158:	sta     $FF08
    000286r 1               ;
    000286r 1               ; port = (*(unsigned char*)0xff08);
    000286r 1               ;
    000286r 1  8D rr rr     	sta     L0118
    000289r 1               ;

This is the workaround:

    000289r 1               ; (*(unsigned char*)0xff08) = 0x04;
    000289r 1               ;
    000289r 1  8D 08 FF     	sta     $FF08
    00028Cr 1               ;
    00028Cr 1               ; port = 0;
    00028Cr 1               ;
    00028Cr 1  8C rr rr     	sty     L0118
    00028Fr 1               ;
    00028Fr 1               ; port = (*(unsigned char*)0xff08);
    00028Fr 1               ;
    00028Fr 1  AD 08 FF     	lda     $FF08 ; <--- this is the step needed
    000292r 1  8D rr rr     	sta     L0118

Command line used to compile:

    cl65 -T -l obj/plus4/midres_plus4.asm -t plus4 -c -W -const-comparison -Osir -Cl -D__CBM__ -o obj/plus4/midres_plus4.o obj/plus4/midres_plus4.c

**Comment by mrdudz** [comment]
this is correct behaviour - the compiler is allowed to do this optimization. to get rid of it, you'd normally use the "volatile" keyword, but cc65 does not support this.

**Comment by Fabrizio-Caruso** [comment]
@mrdudz, clearly having "volatile" would help a lot with most 8-bit targets where some memory locations may be mapped to ports or something else whose value is volatile. I understand this is hard and we have to live with it.
Is there a recommended workaround? pragmas to remove a specific optimization? Special compiler options? Inline Assembly would also fail as it is optimized (I guess).

**Comment by mrdudz** [comment]
I'd put such things into asm code (and asm files) - because even with "volatile" (or without optimizations) the compiler is not supposed to guarantee a certain access pattern.

**Comment by oliverschmidt** [comment]
> I'd put such things into asm code (and asm files) - because even with "volatile" (or without optimizations) the compiler is not supposed to guarantee a certain access pattern.

* At least one of the use cases of a C compiler is to allow for low level programming without asm programming.
* The JSR/RTS combo alone is 12 cycles (not even talking about the optimizer being required to presume most state lost) so putting hardware access into asm functions isn't always a viable option from the performance perspective.
* At least originally `volatile` was explicitly meant to be used for access to memory-mapped I/O devices.

**Comment by oliverschmidt** [comment]
> Is there a recommended workaround? pragmas to remove a specific optimization?

The problem is that you can turn optimization on/off only on function level (using https://cc65.github.io/doc/cc65.html#ss7.11):
* If you put (too) little code into a non-optimized (or asm written) sub-function, then you have (too) much function call overhead both for the actual JSR/RTS instructions and the missed optimization opportunities.
* If you put (too) much code into a non-optimized sub-function, then you simply have (too) much slow code.

**Comment by mrdudz** [comment]
> * At least one of the use cases of a C compiler is to allow for low level programming without asm programming.
> * The JSR/RTS combo alone is 12 cycles (not even talking about the optimizer being required to presume most state lost) so putting hardware access into asm functions isn't always a viable option from the performance perspective.
> * At least originally `volatile` was explicitly meant to be used for access to memory-mapped I/O devices.

That's all fine - if you have a fully working C Compiler =)

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt but do I have an overhead if I keep a small C function in my code and just disable optimization for it?
So, at least for small functions, is the pragma an acceptable workaround?

Ideally it would be very nice to have support for volatile. I suppose it is not easy to implement it as the peephole rules are not applied on the C source code. Are they? So, volatile should be kept somehow and passed to the optimizer to let it know it should not match any pattern and apply any optimization involving a specific location.

**Comment by oliverschmidt** [comment]
To cut a long story short: From my perspective every optimizing C compiler targeting a CPU/MCU with memory-mapped I/O is supposed to have mature `volatile` support - but cc65 doesn't :-(

**Comment by oliverschmidt** [comment]
> but do I have an overhead if I keep a small C function in my code and just disable optimization for it?

That totally depends on the use case. If the missing `volatile` support urges you to replace some indirect assignment with a sub-function call doing the assignment and you have to call that function in a tight loop then you'll loose (too) much performance.

**Comment by Fabrizio-Caruso** [comment]
Maybe there is a way to implement some rudimentary support for volatile.
I am trying to guess how CC65 works. So please forgive me if my idea is not practical:
Could we at least have the "volatile" keyword disable all "access modifying optimizations" on the whole function?
CC65 already has a pragma to disable all optimizations. I suppose it could be possible to disable only optimizations that can change the memory access. Some substitutions are harmless and should be kept.

**Comment by oliverschmidt** [comment]
> That's all fine - if you have a fully working C Compiler =)

So we all agree that this issue is a valid enhancement request - as @greg-king5 already marked :-)

> I am trying to guess how CC65 works. [...]

I don't know how cc65 works. Maybe someone else wants to comment...

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt is my suggestion for a quick and dirty volatile non-sense?
As a workaround, why not have a pragma that works on a single function as the one suggested above but that only disables peephole rules that can destroy volatile variables. I suppose some peephole rules are harmless for volatile variables.
Such a pragma could be used with fewer negative side-effects.
A volatile keyword may be implemented in the same way: it affects the whole function and blocks all dangerous optimizations on the hole function. I understand this may generated sub-optimal code but at least the code would work. Working code is better than broken code.

**Comment by polluks** [comment]
@Fabrizio-Caruso I suppose the effort would be the same for volatile keyword and pragma. What about our two new core developers? I'm pretty sure they are smart enough.

**Comment by spotlessmind1975** [comment]
First of all, thank you very much for the very fast feedback. Above all, for the competence with which you have tagged and dealt with my report and for the comments, very interesting and appropriate. I hope I could give you a useful suggestion as your valuable comments have been for me.

I state that I do not know the internals of the optimizer thoroughly, and therefore I cannot know if my solution is feasible or not. However, I think that the indication of memory volatility is an attribute of the target platform rather than the code, i.e. it should not be handled by the `volatile` attribute.

Let me explain better.

The problem arises from the fact that there are some locations, for a specific platform, that are non-idempotent. That is, the value written in these locations does not mean that the same value will be read, which is the heart of this issue. The set of these locations depends on the target platform: those for the Commodore 64 are different from those for the PLUS/4.

If it were possible to register these locations, and to instruct the optimizer to avoid optimization if the source and/or destination addresses refer to those locations, we would have solved the problem without modifying the original C code. Regardless of whether you access it directly or indirectly. I emphasize that this would imply that a code that does not work today, due to optimization, would work again. A code that already works, on the other hand, will continue to work.

For example, in the case of my code it could be something like:

    // This is my specific case
    // A more general could be defined in the "plus4.h" include file as:
    // #pragma(memory,0xff00-0xff3f)
    #pragma(memory,0xff08)
    
    unsigned char port;
    [...]
    (*(unsigned char*)0xff08) = 0x04;
    port = (*(unsigned char*)0xff08); 

Clearly, the list (or range) of memory locations could be done by the programmer, and a predefined set given along the compiler (with a specific include file platform dependant).

**Comment by Fabrizio-Caruso** [comment]
I think @spotlessmind1975 is making a good point on the need for a pragma because volatile for CC65 seems to be only necessary to pointers to just tell the compiler not to optimize any access to the corresponding memory locations.
I would still recommend to also have volatile support to let the user chose to use standard C idioms.
The suggested pragma could be implemented with the same code and would allow CC65 users to avoid unnecessary volatile assignments in some memory tight situations.

On the other hand, I feat that having a list of such volatile addresses per target cannot be done as such address with either bank-swiching/mappers or reconfigurable ports/registers on some targets may refer to more than one thing.

**Comment by oliverschmidt** [comment]
> A volatile keyword may be implemented in the same way: it affects the whole function and blocks all dangerous optimizations on the hole function.

https://github.com/cc65/cc65/commit/2c7ccca2103c159b27b696dbf263b86eacf4f69c introduced that very behavior for `__asm__`.

> However, I think that the indication of memory volatility is an attribute of the target platform rather than the code, i.e. it should not be handled by the volatile attribute.

I certainly don't think so...

> [...] having a list of such volatile addresses per target cannot be done as such address with either bank-swiching/mappers or reconfigurable ports/registers on some targets may refer to more than one thing.

...because of that.

> [...] be only necessary to pointers to just tell the compiler not to optimize any access to the corresponding memory locations.

That's exactly what `volatile` is about. https://en.wikipedia.org/wiki/Volatile_(computer_programming) says: *Volatile values primarily arise in hardware access (memory-mapped I/O), where reading from or writing to memory is used to communicate with peripheral devices [...]*

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt I agree with everything except on a detail (which is insignificant for CC65).
In ANSI C volatile may also refer to non-pointer variable. The most common use-case for volatile non-pointer values arises with shared variables in multi-threading. This is clearly out of the scope of CC65.

This is why volatile for CC65 (and not for general ANSI C) only has the first use-case you mention. So it is only about addresses (and not about shared variables. Unless one day some implements some sort of shared variables with parallelism through interrupts but I consider this rather far-fetched.

So in the end we only need to flag some addressed as volatile. This could be done with a pragma but I would also like to have the volatile keyboard to let people choose whether to use the standard C syntax.

**Comment by oliverschmidt** [comment]
> In ANSI C volatile may also refer to non-pointer variable. The most common use-case for volatile non-pointer values arises with shared variables in multi-threading. This is clearly out of the scope of CC65.

Now I see your point - thanks for clarifying.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt this is why @spotlessmind1975 's suggestion to have a pragma makes sense.
If we had volatile in CC65, we would only use it to assign it to pointers just to do the very same thing we would do with a pragma. I would still like to have the volatile because people may want to use some idioms or code in a standard way.
The advantage of a pragma is that you avoid the instantiation of a variable whose only purpose is to inform the compiler not to optmize accesses through it.
Moreover volatile in ANSI C is not just  an attribute of the variable. It is an attribute of a type. So it is part of the signature. I fear this make make the life of core devs here harder. I personally would like to have it, though.

**Comment by oliverschmidt** [comment]
Likely there's a misunderstanding. I didn't (want to) comment on the question "pragma and/or volatile". I only wanted to comment on 

> the indication of memory volatility is an attribute of the target platform

where I don't (generally) agree with. However, in case we would have such a pragma I could imagine that there are targets which could benefit from such pragma "predefined" in a target header file.

**Comment by groessler** [comment]
> This is clearly out of the scope of CC65.

Why?

**Comment by Marrin** [comment]
@Fabrizio-Caruso 

> @oliverschmidt I agree with everything you have written except for a detail (which is insignificant for CC65).
> In ANSI C volatile may also refer to non-pointer variable. The most common use-case for volatile non-pointer values arises with shared variables in multi-threading. This is clearly out of the scope of CC65.

It's not shared variables in multi-threading but shared variables in concurrent programming. This includes interrupt routines. And you don't even need to write your own — some target's operating systems use them and store interesting stuff in known memory locations. So this isn't very far fetched.  Also you can define `extern` variables that are not pointers but refer to hardware registers.

**Comment by Fabrizio-Caruso** [comment]
@Marrin if it is memory location than it is not the casa I (maybe wrongly) considered far-fetched. I said that we need volatile addresses (functionally similar to volatile pointer variables) but maybe not volatile non-pointer variables. 
Volatile non-pointer variables are usually used in concurrent programming, which I considered (probably wrongly) a far-fetched use-case.

**Comment by Fabrizio-Caruso** [comment]
@oliverschmidt, Support for volatile may mean a lot of work on the core devs. Am I wrong?
As a temporary solution, could the core devs implement a variation of the currently available pragma to just disable access-modifying optimizations on a function? Would it make sense?

**Comment by oliverschmidt** [comment]
> As a temporary solution, could the core devs implement a variation of the currently available pragma to just disable access-modifying optimizations on a function? Would it make sense?

I can for sure follow your thoughts, but I as I'm none of the core devs I can't assess if they actually make sense.

However, you might be interested to "play" with the current code base. If you put the function(s) in question into a separate source file you can have finer control over optimization. Check out the following compiler cmdline options:

* --list-opt-steps
* --enable-opt
* --disable-opt

If that approach allows you to generate code both fast (enough) and "volatile compatible" enough, it shouldn't be too hard to create a pragma supporting the same optimizer settings on a per-function level.

**Comment by acqn** [comment]
> To cut a long story short: From my perspective every optimizing C compiler targeting a CPU/MCU with memory-mapped I/O is supposed to have mature `volatile` support - but cc65 doesn't :-(

I've got `volatile` support in the front end (parser+"AST"-ish ExprDesc IR) in my IR branch (aside from any hidden bugs if there are). Unfortunately, it isn't easy to even make the back end (assembler+peephole-optimizer) fully aware of and respect it.

But wait, there are questions about supporting `volatile` as the way to deal with memory-mapped I/O registers:
```c
extern int a;
extern volatile int v, w;
extern volatile struct Z1 { int a : 1; int b : 6; int c : 1; } z1;
extern volatile struct Z2 { int a : 2; int b : 3; int c : 2; } z2;
v = a;        /* Does the order of Writing to the two bytes matter? */
a = v;        /* Does the order of Reading from the two bytes matter? */
++v;          /* Can we interlace R/W of the two bytes? */
w = v++;      /* What is the "correct" order to do the R/W? */
z1.b += z2.b; /* How should we do all the R/W? */
```
Yeah there are existing C compilers dealing with these but we need decide our answers (even if they were "just do what xxx does").

**Comment by mrdudz** [comment]
IMHO for reads or writes that are not atomic (ie char types) the user is responsible to break it up as needed. I remember that being the case on other platforms/compilers i have used too (keill perhaps, or sdcc, i don't recall)

**Comment by Fabrizio-Caruso** [comment]
Any progress on this feature request ? Even partial sub-optimal volatile support would help IMHO.

**Comment by mrdudz** [comment]
I don't think anyone is working on it (and @acqn is MIA unfortunately)

**Comment by safiire** [comment]
Having the compiler emit a warning if you use `volatile`, letting you know it isn't implemented would be helpful, and reduce the frustration of wondering what you did wrong until you look on github issues.
